src/main/scala/mesosphere/marathon/core/task/termination/impl/TaskKillServiceActor.scala

```diff
@@ -0,0 +1,183 @@
+package mesosphere.marathon.core.task.termination.impl
+
+import akka.Done
+import akka.actor.{ Actor, ActorLogging, Cancellable, Props }
+import mesosphere.marathon.MarathonSchedulerDriverHolder
+import mesosphere.marathon.core.base.Clock
+import mesosphere.marathon.core.task.termination.TaskKillConfig
+import mesosphere.marathon.state.Timestamp
+import mesosphere.marathon.core.task.{ Task, TaskStateOp }
+import mesosphere.marathon.core.task.tracker.{ TaskStateOpProcessor, TaskTracker }
+import mesosphere.marathon.core.event.MesosStatusUpdateEvent
+
+import scala.collection.mutable
+import scala.concurrent.Promise
+
+/**
+  * An actor that handles killing tasks in chunks and depending on the task state.
+  * Lost tasks will simply be expunged from state, while active tasks will be killed
+  * via the scheduler driver. There is be a maximum number of kills in flight, and
+  * the service will only issue more kills when tasks are reported terminal.
+  *
+  * If a kill is not acknowledged with a terminal status update within a configurable
+  * time window, the kill is retried a configurable number of times. If the maximum
+  * number of retries is exceeded, the task will be expunged from state similar to a
+  * lost task.
+  *
+  * For each kill request, a child [[TaskKillProgressActor]] will be spawned, which
+  * is supposed to watch the progress and complete a given promise when all watched
+  * tasks are reportedly terminal.
+  *
+  * See [[TaskKillConfig]] for configuration options.
+  */
+private[impl] class TaskKillServiceActor(
+    taskTracker: TaskTracker,
+    driverHolder: MarathonSchedulerDriverHolder,
+    stateOpProcessor: TaskStateOpProcessor,
+    config: TaskKillConfig,
+    clock: Clock) extends Actor with ActorLogging {
+  import TaskKillServiceActor._
+  import context.dispatcher
+
+  val tasksToKill: mutable.HashMap[Task.Id, Option[Task]] = mutable.HashMap.empty
+  val inFlight: mutable.HashMap[Task.Id, TaskToKill] = mutable.HashMap.empty
+  var retryTimer: Option[Cancellable] = None
+
+  override def preStart(): Unit = {
+    context.system.eventStream.subscribe(self, classOf[MesosStatusUpdateEvent])
+  }
```

Comment:
add postStop:
- unsubscribe from the event stream.
- cancel the retryTimer
- if not all tasks were killed, log the remaining tasks